File: Behavioral/Chain of responsibility/Delphi/BaseHandler.pas
unit BaseHandler;

interface

uses
  Handler;

type
  TBaseHandler = class(TInterfacedObject, IHandler)
  private
    FNextHandler: IHandler;
  public
    constructor Create();
    destructor Destroy(); override;
    procedure Handle(request: string); virtual;
    function SetNext(handler: IHandler): IHandler;
  end;

implementation

constructor TBaseHandler.Create();
begin
  inherited Create();
end;

destructor TBaseHandler.Destroy();
begin
  if Assigned(FNextHandler) then
    FNextHandler := nil;
  inherited Destroy();
end;

function TBaseHandler.SetNext(handler: IHandler): IHandler;
begin
  Result := handler;
  FNextHandler := handler;
end;

procedure TBaseHandler.Handle(request: string);
begin
  if Assigned(FNextHandler) then
    FNextHandler.Handle(request);
end;

end.

File: Behavioral/Chain of responsibility/C#/ChainOfResponsibilityExample/ChainOfResponsibilityExample/BaseHandler.cs
namespace ChainOfResponsibilityExample;

abstract class BaseHandler : IRequestHandler
{
	protected IRequestHandler? NextHandler { get; set; }

	public virtual void Handle(string request)
	{
		if (NextHandler != null)
			NextHandler.Handle(request);
	}

	public IRequestHandler SetNext(IRequestHandler next)
	{
		this.NextHandler = next;
		return next;
	}
}